File: e_zyro_app/lib/services/usuarios_service.dart
import 'dart:convert';
import '../core/api_client.dart';
import '../core/api_result.dart';
import '../models/usuario_admin_models.dart';

/// Cliente de Gestión de usuarios (`/usuarios`). Solo Admin / permiso
/// `usuarios:gestionar`. Todas las operaciones quedan acotadas a la empresa
/// del token en el backend.
class UsuariosService {
  final ApiClient _client;
  UsuariosService(this._client);

  Future<ApiResult<List<UsuarioAdmin>>> listar() async {
    try {
      final r = await _client.get('/usuarios');
      if (r.statusCode == 200) {
        final list = jsonDecode(r.body) as List;
        return ApiResult.ok(
            list.map((e) => UsuarioAdmin.fromJson(e as Map<String, dynamic>)).toList());
      }
      return ApiResult.fail(ApiError.fromResponse(r));
    } catch (_) {
      return const ApiResult.fail(ApiError(ApiErrorKind.network));
    }
  }

  Future<ApiResult<List<RolItem>>> roles() async {
    try {
      final r = await _client.get('/usuarios/roles');
      if (r.statusCode == 200) {
        final list = jsonDecode(r.body) as List;
        return ApiResult.ok(
            list.map((e) => RolItem.fromJson(e as Map<String, dynamic>)).toList());
      }
      return ApiResult.fail(ApiError.fromResponse(r));
    } catch (_) {
      return const ApiResult.fail(ApiError(ApiErrorKind.network));
    }
  }

  Future<ApiResult<UsuarioAdmin>> crear({
    required String nombre,
    required String apellido,
    required String email,
    required String username,
    required String password,
    required String rolId,
    bool crearFicha = false,
    String? cargo,
    String? area,
    String? tipo,
    String? codigo,
  }) async {
    try {
      final r = await _client.post('/usuarios', {
        'nombre': nombre,
        'apellido': apellido,
        'email': email,
        'username': username,
        'password': password,
        'rol_id': rolId,
        'crear_ficha': crearFicha,
        'cargo': ?cargo,
        'area': ?area,
        'tipo': ?tipo,
        'codigo': ?codigo,
      });
      if (r.statusCode == 201 || r.statusCode == 200) {
        return ApiResult.ok(UsuarioAdmin.fromJson(jsonDecode(r.body) as Map<String, dynamic>));
      }
      return ApiResult.fail(ApiError.fromResponse(r));
    } catch (_) {
      return const ApiResult.fail(ApiError(ApiErrorKind.network));
    }
  }

  Future<ApiResult<UsuarioAdmin>> cambiarRol(String usuarioId, String rolId) async {
    try {
      final r = await _client.put('/usuarios/$usuarioId/rol', {'rol_id': rolId});
      if (r.statusCode == 200) {
        return ApiResult.ok(UsuarioAdmin.fromJson(jsonDecode(r.body) as Map<String, dynamic>));
      }
      return ApiResult.fail(ApiError.fromResponse(r));
    } catch (_) {
      return const ApiResult.fail(ApiError(ApiErrorKind.network));
    }
  }

  Future<ApiResult<UsuarioAdmin>> cambiarActivo(String usuarioId, bool activo) async {
    try {
      final r = await _client.put('/usuarios/$usuarioId/activo', {'activo': activo});
      if (r.statusCode == 200) {
        return ApiResult.ok(UsuarioAdmin.fromJson(jsonDecode(r.body) as Map<String, dynamic>));
      }
      return ApiResult.fail(ApiError.fromResponse(r));
    } catch (_) {
      return const ApiResult.fail(ApiError(ApiErrorKind.network));
    }
  }

  Future<ApiResult<void>> resetPassword(String usuarioId, String password) async {
    try {
      final r = await _client.post('/usuarios/$usuarioId/reset-password', {'password': password});
      if (r.statusCode == 200) return const ApiResult.ok();
      return ApiResult.fail(ApiError.fromResponse(r));
    } catch (_) {
      return const ApiResult.fail(ApiError(ApiErrorKind.network));
    }
  }
}

File: src/app/core/services/usuarios.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

export interface UsuarioOut {
  id: string;
  nombre: string;
  apellido: string;
  email: string;
  username: string;
  rol: string | null;
  rol_id: string | null;
  activo: boolean;
  tiene_ficha: boolean;
}

export interface RolOut {
  id: string;
  nombre: string;
}

export interface CrearUsuarioIn {
  nombre: string;
  apellido: string;
  email: string;
  username: string;
  password: string;
  rol_id: string;
  telefono?: string;
  crear_ficha: boolean;
  cargo?: string;
  area?: string;
  tipo?: string;
  codigo?: string;
  tipo_documento?: string;
  numero_documento?: string;
  sexo?: string;
  fecha_ingreso?: string;
}

@Injectable({ providedIn: 'root' })
export class UsuariosService {
  private http = inject(HttpClient);
  private api  = environment.apiUrl;

  listar(): Observable<UsuarioOut[]> {
    return this.http.get<UsuarioOut[]>(`${this.api}/usuarios`);
  }

  getRoles(): Observable<RolOut[]> {
    return this.http.get<RolOut[]>(`${this.api}/usuarios/roles`);
  }

  crear(body: CrearUsuarioIn): Observable<UsuarioOut> {
    return this.http.post<UsuarioOut>(`${this.api}/usuarios`, body);
  }

  cambiarRol(id: string, rol_id: string): Observable<UsuarioOut> {
    return this.http.put<UsuarioOut>(`${this.api}/usuarios/${id}/rol`, { rol_id });
  }

  cambiarActivo(id: string, activo: boolean): Observable<UsuarioOut> {
    return this.http.put<UsuarioOut>(`${this.api}/usuarios/${id}/activo`, { activo });
  }

  resetPassword(id: string, password: string): Observable<{ status: string; mensaje: string }> {
    return this.http.post<{ status: string; mensaje: string }>(`${this.api}/usuarios/${id}/reset-password`, { password });
  }
}
